### PR TITLE
Added std::invoke_result for C⁺⁺17

### PR DESCRIPTION
**Pull request by Jamaika1**

### CODE REVIEW

File: mingw.invoke.h [hunk 1]
```diff
@@ -21,27 +21,35 @@ namespace mingw_stdthread
...
   {
     template<class F, class... Args>
+#if __cplusplus >= 201703L
+    inline static typename std::invoke_result<F, Args...>::type invoke (F&& f, Args&&... args)
+#else
     inline static typename std::result_of<F(Args...)>::type invoke (F&& f, Args&&... args)
+#endif
     {
+#if __cplusplus >= 201703L
+      return std::__invoke(std::forward<F>(f), std::forward<Args>(args)...);
```
Comment (nmcclatchey) [inline mingw.invoke.h:35]: It's better not to rely on the current library implementation, and instead rely on [the standard](https://en.cppreference.com/w/cpp/utility/functional/invoke).

File: mingw.invoke.h [hunk 2]
```diff
@@ -89,20 +98,19 @@ namespace detail
...
 
   template<class F, class...Args>
-  auto invoke (F&& f, Args&&... args) -> decltype(InvokeResult<F, Args...>::invoke(std::forward<F>(f), std::forward<Args>(args)...))
+  auto invoke (F&& f, Args&&... args) ->\
+   decltype(InvokeResult<F, Args...>::invoke(std::forward<F>(f), std::forward<Args>(args)...))
   {
     return InvokeResult<F, Args...>::invoke(std::forward<F>(f), std::forward<Args>(args)...);
   }
-#else
-    using std::invoke;
```
Comment (nmcclatchey) [inline mingw.invoke.h:104]: Removing this subverts the entire intent of your pull request.

Before you removed this, the code would use the C++17 standard's `std::invoke` when compiling with C++17, and use a fallback implementation when compiling with C++11 or C++14. With your changes, `std::invoke` is only used as a fallback in an implementation of `std::invoke`; this can only add inefficiency.

File: mingw.invoke.h [hunk 3]
```diff
@@ -21,27 +21,35 @@ namespace mingw_stdthread
...
   template<>
   struct InvokerHelper<false>
   {
     template<class T1>
-    inline static auto get (T1&& t1) -> decltype(*std::forward<T1>(t1))
+    inline static auto get (T1&& t1) ->\
+     decltype(std::forward<T1>(t1))
     {
-      return *std::forward<T1>(t1);
+      return std::forward<T1>(t1);
```
Comment (nmcclatchey) [inline mingw.invoke.h:52]: Your change here appears to violate ["If t1 does not satisfy the previous items, then INVOKE(f, t1, t2, ..., tN) is equivalent to ((*t1).*f)(t2, ..., tN)."](https://en.cppreference.com/w/cpp/utility/functional/invoke) from the C++ standard.

**Review by nmcclatchey — CHANGES_REQUESTED** [review]
This pull request is redundant. If compiling with C++17, `std::invoke` (and implicitly `std::invoke_result`) are already used. Additionally, the implementation appears to introduce at least one bug.

Please state what problem is intended to be solved by this pull request, and see the comments left as part of this review.

File: mingw.invoke.h (hunk 1 above)
Comment (Jamaika1) [reply mingw.invoke.h:35]: On the standart std::result_of is removed in C++20.
The latest htj2k codecs are already C++20.

File: mingw.invoke.h (hunk 3 above)
Comment (Jamaika1) [reply mingw.invoke.h:52]: That's why line 66 is different for true.

File: mingw.invoke.h (hunk 1 above)
Comment (Jamaika1) [reply mingw.invoke.h:35]: I was thinking about Invoke a callable object in gcc 12.0.0
```
#if __cplusplus >= 201703L
# define __cpp_lib_invoke 201411L
# if __cplusplus > 201703L
#  define __cpp_lib_constexpr_functional 201907L
# endif

  /// Invoke a callable object.
  template<typename _Callable, typename... _Args>
    inline _GLIBCXX20_CONSTEXPR invoke_result_t<_Callable, _Args...>
    invoke(_Callable&& __fn, _Args&&... __args)
    noexcept(is_nothrow_invocable_v<_Callable, _Args...>)
    {
      return std::__invoke(std::forward<_Callable>(__fn),
			   std::forward<_Args>(__args)...);
    }

#if __cplusplus > 202002L
# define __cpp_lib_invoke_r 202106L

  /// Invoke a callable object and convert the result to _Res.
  template<typename _Res, typename _Callable, typename... _Args>
    requires is_invocable_r_v<_Res, _Callable, _Args...>
    constexpr _Res
    invoke_r(_Callable&& __fn, _Args&&... __args)
    noexcept(is_nothrow_invocable_r_v<_Res, _Callable, _Args...>)
    {
      return std::__invoke_r<_Res>(std::forward<_Callable>(__fn),
				   std::forward<_Args>(__args)...);
    }
#endif // C++23
#endif // C++17
 ```

Comment (nmcclatchey) [reply mingw.invoke.h:35]: This is already handled in the existing library code. Are you seeing errors during compilation?

If you had not removed the line `#if __cplusplus < 201703L` and its corresponding lines
```
#else
using std::invoke;
#endif
```
Then the compatibility implementation would not be selected when compiling in C++20 (and, in fact, would not be compiled at all). Instead, the line `using std::invoke;` would pull in the standard library's `invoke` implementation, which then uses `invoke_result` to return type.

In short, you intend to add C++20 compatibility to code that is `#ifdef`'ed away for C++17 and higher.

Comment (Jamaika1) [reply mingw.invoke.h:35]: I am an amateur of programming. So out of boredom he asks questions.

File: mingw.invoke.h (hunk 3 above)
Comment (nmcclatchey) [reply mingw.invoke.h:52]: The change you submitted in the line that I highlighted will break the case of `invoke(pointer_to_data_member, pointer_to_object)`. The change on line 66 does not help, due to being a pure whitespace change.

To borrow and modify a small piece of the example from [cppreference.com](https://en.cppreference.com/w/cpp/utility/functional/invoke), the standard indicates that we can write:
```
struct Foo {
    int num_ = 314159;
};
 
int main()
{ 
    using namespace mingw_stdthread::detail;
    // invoke a member function
    const Foo foo(314159);
 
    // invoke (access) a data member using a pointer to the object.
    std::cout << "num_: " << invoke(&Foo::num_, &foo) << '\n';
}
```

This should compile without issues in the base mingw-std-threads library (with appropriate header inclusions, of course). However, I expect your modification to cause a compilation error.

Comment (Jamaika1) [reply mingw.invoke.h:52]: Thanks for the explanations. I already know that std :: result_of is complemented with invoke_result functions. :)
